src/steamship/agents/agents.py

```diff
@@ -0,0 +1,187 @@
+import uuid
+from abc import ABC, abstractmethod
+from typing import Any, List, Optional
+
+from pydantic import BaseModel
+
+from steamship import Block, File, Steamship, Tag, Task
+from steamship.invocable import InvocableResponse, PackageService, post
+from steamship.utils.kv_store import KeyValueStore
+
+
+class Action(BaseModel):
+    name: str
+    data: Any
+
+
+class FinishAction(Action):
+    pass
+
+
+class AgentContext(BaseModel, ABC):
+    # todo: define what should be in this context
+    # some set of state-related to agent (maybe request ids, callbacks, etc.)
+    pass
+
+    @abstractmethod
+    def update_blocks(self, blocks: List[Block]):
+        pass
+
+    @abstractmethod
+    def append_log(self, message: str):
+        pass
+
+
+class Tool(BaseModel, ABC):
+    name: str
+    human_description: str
+    ai_description: str
+
+    @abstractmethod
+    def run(self, tool_input: List[Block], context: AgentContext) -> List[Block]:
```

Comment:
i wonder if we want wrapper base methods that do things like report start/stop, etc.

Reply:
Yeah -- and I think callbacks into the agent or python too.

Since a lot of times the user of a tool might not be the developer of the tool (or in the ReACt case.. might not even be the direct invoker!)

So that the wrapping agent or even packageService could know when a tool was about to execute, did execute, etc.